contracts/Governance/Timelock.sol

```diff
@@ -1,123 +1,160 @@
-        require(queuedTransactions[txHash], "Timelock::executeTransaction: Transaction hasn't been queued.");
-        require(getBlockTimestamp() >= eta, "Timelock::executeTransaction: Transaction hasn't surpassed time lock.");
-        require(getBlockTimestamp() <= eta + GRACE_PERIOD, "Timelock::executeTransaction: Transaction is stale.");
+        _approveTransactionsBatchInternal(txParamsHash);
+        _executeTransactionsBatchInternal(transactions, eta);
+    }
 
-        queuedTransactions[txHash] = false;
+    function _approveTransactionsBatchInternal(bytes32 txParamsHash) internal onlyOwner {
+        require(queuedTransactions[txParamsHash] == TrnasactionStatus.Queued, "Timelock: transaction is not queued");
 
-        bytes memory callData;
+        queuedTransactions[txParamsHash] = TrnasactionStatus.Approved;
 
-        if (bytes(signature).length == 0) {
-            callData = data;
-        } else {
-            callData = abi.encodePacked(bytes4(keccak256(bytes(signature))), data);
-        }
+        emit ApprovedTransactionsBatch(txParamsHash);
+    }
+
+    function _executeTransactionsBatchInternal(Transaction[] memory transactions, uint256 eta) internal {
+        bytes32 txParamsHash = keccak256(abi.encode(transactions, eta));
 
-        // Execute the call
-        (bool success, bytes memory returnData) = target.call{value: value}(callData);
-        require(success, "Timelock::executeTransaction: Transaction execution reverted.");
+        require(queuedTransactions[txParamsHash] == TrnasactionStatus.Approved, "Timelock: Transaction hasn't been approved.");
+        require(block.timestamp >= eta, "Timelock: Transaction hasn't surpassed time lock.");
+        require(block.timestamp <= eta + gracePeriod, "Timelock: Transaction is stale.");
 
-        emit ExecuteTransaction(txHash, target, value, signature, data, eta);
+        delete queuedTransactions[txParamsHash];
 
-        return returnData;
+        for (uint256 i = 0; i < transactions.length; i++) {
+            bytes memory callData = bytes(transactions[i].signature).length == 0
+                ? transactions[i].data
+                : abi.encodePacked(bytes4(keccak256(bytes(transactions[i].signature))), transactions[i].data);
+
+            // Execute the transaction
+            (
+                bool success, /* bytes memory returnData */
+
+            ) = transactions[i].target.call{value: transactions[i].value}(callData);
+            require(success, "Timelock: Transaction execution reverted.");
+
+            emit ExecutedTransaction(
+                txParamsHash,
+                transactions[i].target,
+                transactions[i].value,
+                transactions[i].signature,
+                transactions[i].data,
+                eta
+            );
+        }
     }
 
-    function getBlockTimestamp() internal view returns (uint256) {
-        return block.timestamp;
+    modifier onlyProposerOrOwner() {
```

Comment:
Why do we need this modifier? The way I see it we should only have a `onlyProposer` modifier.. Is there a real need for the owner to be able to run tasks meant for the proposer? If so, where?

The owner could always change the proposer if needed..